posts/simprocs-intro.md

```diff
@@ -0,0 +1,550 @@
+* Computation
+
+### Avoiding combinatorial explosion of lemmas: The `existsAndEq` simproc
+
+The `existsAndEq` simproc is designed to simplify expressions of the form `∃ a, ... ∧ a = a' ∧ ...` where `a'` is some quantity independent of `a'` by removing the existential quantifier and replacing all occurences of `a` by `a'`.
+
+```lean
+example : ∃ (a : ℤ), a*a = 25 ∧ a = 5 := by
+  simp +arith only [existsAndEq, and_true]
+
+example : (∃ a, (∃ b, a + b = 5) ∧ a = 3) ↔ ∃ b, 3 + b = 5 := by
+  simp only [existsAndEq, and_true]
+```
+
+Roughly speaking, the way this metaprogram operates is a follows: whenever an expression of the form `∃ a, P a` with `P a = Q ∧ R`  is encountered:
+- Recursively traverse the expression of the predicate inside the existential quantifier to try and detect an equality `a = a'` by splitting any `∧` that is found along the way into its components.
+- If an equality is found, construct a proof that `P a` implies `a = a`.
+- Construct a proof that `(∃ a, P a) = P a'` using the following theorem
+```lean
+theorem exists_of_imp_eq {α : Sort u} {p : α → Prop} (a : α) (h : ∀ b, p b → a = b) :
+    (∃ b, p b) = p a
+```
+
+### Performance optimisation: The `reduceIte` simproc
+
+The `reduceIte` simproc is designed to take expressions of the form `ite P a b` and replace them with `a` or `b`, depending on whether `P` can be simplified to `True` or `False` by a `simp` call.
+
+```lean
+example : ite (1 + 1 = 2) 1 2 = 1 := by
+  -- Works since `simp` can simplify `1 + 1 = 2` to `True`.
+  simp only [↓reduceIte]
+
+example (T : Type) (P : Prop) (a b : T) : ite (P ∨ ¬ P) a b = a := by
+  -- Works since `simp` can simplify `P ∨ ¬ P` to `True`.
+  simp only [↓reduceIte]
+
+example : ite FermatLastTheorem 1 2 = 1 := by
+  --This fails because `simp` can't reduce `FermatLastTheorem` to `True`
+  simp only [↓reduceIte]
+```
+
+Internally, this simproc is a small metaprogram does the following whenever an expression of the form `ite P a b` is encountered:
+- Call `simp` on `P` to get a simplified expression `P'` and a proof `r` that `P = P'`.
+**TODO Paul** check that the arguments in the proof are alright.
+- If `P' = True` then return the simplified expression `a` and the proof `ite_cond_eq_true r` that `ite P a b = a`
+- If `P' = False` then return the simplified expression `b` and the proof `ite_cond_eq_false r` that `ite P a b = b`
+- Otherwise, let `simp` continue the search.
+
+Source code is
+```lean
+builtin_simproc ↓ [simp, seval] reduceIte (ite _ _ _) := fun e => do
+  let_expr f@ite α c i tb eb ← e | return .continue
+  let r ← simp c
+  if r.expr.isTrue then
+    let pr    := mkApp (mkApp5 (mkConst ``ite_cond_eq_true f.constLevels!) α c i tb eb) (← r.getProof)
+    return .visit { expr := tb, proof? := pr }
+  if r.expr.isFalse then
+    let pr    := mkApp (mkApp5 (mkConst ``ite_cond_eq_false f.constLevels!) α c i tb eb) (← r.getProof)
+    return .visit { expr := eb, proof? := pr }
+  return .continue
+```
```

Comment:
This is really intimidating. Do you really want this here, so early? Until now the post has been easy-going. It's fine to have "accelerated lecture style" with code like this later on, but the sooner you put some blob of incomprehensible code the sooner you lose readers. If you're trying to tell us something with this code, can you tell it us without any meta code at all? You claimed earlier that the first section was purely informal.

Reply:
Could be a nice compromise to just link to the documentation for `reduceIte` and suggest pressing "source" for the interested user.

Reply:
All gone now